src/circuit/schedule/mod.rs

```diff
@@ -130,3 +142,109 @@ where
         self.scheduler.step(circuit)
     }
 }
+
+mod util {
+    //! Some useful tools for developing schedulers.
+
+    use crate::circuit::{schedule::Error, Circuit, GlobalNodeId, NodeId, OwnershipPreference};
+    use petgraph::graphmap::DiGraphMap;
+    use std::{collections::HashMap, ops::Deref};
+
+    /// Dump circuit topology as a graph.
+    pub(crate) fn circuit_graph<P>(circuit: &Circuit<P>) -> DiGraphMap<NodeId, ()> {
+        let mut g = DiGraphMap::<NodeId, ()>::new();
+
+        for node_id in circuit.node_ids().into_iter() {
+            g.add_node(node_id);
+        }
+
+        for edge in circuit.edges().deref().iter() {
+            if let Some(from) = edge.from {
+                g.add_edge(from, edge.to, ());
+            }
+        }
+
+        g
+    }
+
+    /// Helper function used by schedulers to enforce ownership preferences.
+    ///
+    /// Individual schedulers can implement their own algorithms to enforce (or ignore)
+    /// ownership preferences.  This helper function can optionally used by schedulers that
+    /// wish to implement one particular approach.  The idea is to treat **strong** ownership
+    /// preferences (`OwnershipPreference::require_owned` and above) as hard constraints:
+    /// assuming an operator has exactly one successor with a strong ownership preference
+    /// ("strong successor"), we can enforce this preference by scheduling this successor last.
+    /// This scheduling constraint can in turn be enforced by adding a dependency edge from
+    /// all other successors to the strong successor node to the circuit graph.  This function
+    /// computes the set of dependency edges needed to enforce all such constraints in the
+    /// circuit.
+    ///
+    /// # Caveat
+    ///
+    /// The additional edges computed by this function can introduce cycles to the circuit
+    /// graph making it unschedulable.  Schedulers must check for cycles before adding these
+    /// constraints and either fail or drop some of the constraints to eliminate cycles.
```

Comment:
I assume that hard constraints cannot be dropped

Reply:
I'll clarify.  The current implementation models them as hard constraints but a different implementation may choose to relax them.